testdata/generate.go

```diff
@@ -0,0 +1,235 @@
+// This binary creates all the necessary certificates and private keys required for the Bootz emulator.
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"math/big"
+	"os"
+	"strings"
+	"time"
+
+	log "github.com/golang/glog"
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	vendor             = flag.String("vendor", "", "The name of the vendor to generate self-signed certificates for.")
+	owner              = flag.String("owner", "", "The name of the organization that owns the emulated device.")
+	controlCardSerials = flag.String("serials", "", "Comma-separated list of control card serials to generate OVs for.")
+)
+
+type OwnershipVoucher struct {
+	OV OwnershipVoucherInner `json:"ietf-voucher:voucher"`
+}
+
+// Defines the Ownership Voucher format. See https://www.rfc-editor.org/rfc/rfc8366.html.
+type OwnershipVoucherInner struct {
+	CreatedOn                  string `json:"created-on"`
+	ExpiresOn                  string `json:"expires-on"`
+	SerialNumber               string `json:"serial-number"`
+	Assertion                  string `json:"assertion"`
+	PinnedDomainCert           string `json:"pinned-domain-cert"`
+	DomainCertRevocationChecks bool   `json:"domain-cert-revocation-checks"`
+}
+
+// certFromPem converts a PEM-formatted byte slice to an x509 Certificate
+func certFromPem(contents []byte) (*x509.Certificate, error) {
+	pemBlock, _ := pem.Decode(contents)
+	return x509.ParseCertificate(pemBlock.Bytes)
+}
+
+// privateKeyFromPem converts a PEM-formatted byte slice to an RSA Private Key
+func privateKeyFromPem(contents []byte) (*rsa.PrivateKey, error) {
+	pemBlock, _ := pem.Decode(contents)
+	return x509.ParsePKCS1PrivateKey(pemBlock.Bytes)
+}
+
+// newCertificateAuthority creates a new CA for the chosen organization.
+// It returns a self-signed CA certificate as the first value, the associated private key as the second and any error as the third.
+func newCertificateAuthority(commonName string, org string) (string, string, error) {
+	// Create the certificate authority.
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(int64(time.Now().Year())),
+		Subject: pkix.Name{
+			CommonName:   commonName,
+			Organization: []string{org},
+			Country:      []string{"US"},
+			Province:     []string{"CA"},
+			Locality:     []string{"Mountain View"},
```

Comment:
These constants will work anyways, but I think we should move them to constants for now and later to move them all to a single constants package or .in file.